jax/core.py

```diff
@@ -2022,34 +2022,38 @@ def check_jaxpr(jaxpr: Jaxpr):
   Raises `JaxprTypeError` if `jaxpr` is determined invalid. Returns `None`
   otherwise.
   """
+  ctx = JaxprPpContext()
+  try: pp_jaxpr(jaxpr, ctx)  # side-effect on ctx, build variable names
+  except: pass
+
   try:
-    _check_jaxpr(jaxpr, [v.aval for v in jaxpr.invars])
+    _check_jaxpr(ctx, jaxpr, [v.aval for v in jaxpr.invars])
   except JaxprTypeError as e:
     if len(e.args) == 2:
-      msg, eqn_idx = e.args
-      jaxpr_str = str(pp_jaxpr_eqn_range(jaxpr, eqn_idx - 10, eqn_idx + 10,
-                                         JaxprPpContext()))
+      msg, eqnidx = e.args
+      jaxpr_str = str(pp_jaxpr_eqn_range(jaxpr, eqnidx - 10, eqnidx + 10, ctx))
     else:
       msg, = e.args
-      jaxpr_str = str(pp_jaxpr_eqn_range(jaxpr, 0, 20, JaxprPpContext()))
+      jaxpr_str = str(pp_jaxpr_eqn_range(jaxpr, 0, 20, ctx))
     msg = "\n\n".join([msg, "while checking jaxpr:", jaxpr_str])
     raise JaxprTypeError(msg) from None
 
-def _check_jaxpr(jaxpr: Jaxpr, in_avals: Sequence[AbstractValue]):
+def _check_jaxpr(ctx: 'JaxprPpContext', jaxpr: Jaxpr,
+                 in_avals: Sequence[AbstractValue]) -> List[AbstractValue]:
 
   def read(v: Atom) -> AbstractValue:
     if isinstance(v, Literal):
       return raise_to_shaped(get_aval(v.val))
     else:
-      typecheck_assert(v in env, f"Variable '{v}' not defined")
+      typecheck_assert(v in env, f"Variable '{pp_var(v, ctx)}' not defined")
       return env[v]
 
   def write(v: Var, a: AbstractValue) -> None:
-    typecheck_assert(v not in env, f"Variable '{v}' already bound")
+    typecheck_assert(v not in env, f"Variable '{pp_var(v, ctx)}' already bound")
     if not isinstance(v, DropVar):
       typecheck_assert(typecompat(v.aval, a),
-                       f"Variable '{v}' inconsistently typed as {a}, "
-                       f"bound as {v.aval}")
+                       f"Variable '{pp_var(v, ctx)}' inconsistently typed as "
+                       f"{pp_aval(a, ctx)}, bound as {pp_aval(v.aval, ctx)}")
```

Comment:
This change is useful for jaxprs involving dynamic shapes, which have variables in types (and hence need ctx to print the correct variable names in types).